cpp/nvtx3.hpp

```diff
@@ -15,21 +15,67 @@
  */
 #pragma once
 
-#if defined(NVTX3_MINOR_VERSION) and NVTX3_MINOR_VERSION < 0
-#error \
-  "Trying to #include NVTX version 3 in a source file where an older NVTX version has already been included.  If you are not directly using NVTX (the NVIDIA Tools Extension library), you are getting this error because libraries you are using have included different versions of NVTX.  Suggested solutions are: (1) reorder #includes so the newest NVTX version is included first, (2) avoid using the conflicting libraries in the same .c/.cpp file, or (3) update the library using the older NVTX version to use the newer version instead."
-#endif
-
+/* Temporary helper #defines, #undef'ed at end of header */
+#define NVTX3_CPP_VERSION_MAJOR 1
```

Comment:
Do we want to keep the `NVTX3` name if we're starting the major version off at 1? These seem like a potential source of confusion if they're not consistent. Instead we could just go with `NVTX`? If not, do we intend to change to `nvtx4` when that comes out?

There's a few different moving pieces wrt to versioning between NVTX C and NVTX C++ and we should try and keep them as consistent as possible.

Reply:
You could follow what the NVTX extensions are doing with calling it compatID to version the extension separately but not make people as confused over all the versions.  CompatID lets me pick marketing hubbub totally separate from code.  Yes v4 likely could warrant its own set of C++ wrappers that can live SxS with v3 C++ wrapper.  This way people can have v3 wrappers optimal for v3 and support v3 compatible tools while v4 requires upgraded tools and could have more efficient ways to do the same thing not possible with v3.

Reply:
I think v4 will end up looking quite a bit different from v3, so I like keeping the C++ wrappers for v3 "scoped" to that version.  It certainly keeps us safe from conflicts in the future.  Even for these temporary macros which are unversioned with respect to the C++ wrappers (i.e. no "_V1" in the macro name) because I just #undef them at the end of the header, if I changed e.g. "NVTX3_CONCAT" to "NVTX_CONCAT", we'd have to be careful in the future of both C and C++ NVTX impls to not use that macro name, even temporarily.  It's annoying that you can't "push" and "pop" #defines, so they don't conflict with pre-existing definitions.  So just being careful to namespace them is a good seatbelt there.

For v4 I am fairly convinced we shouldn't have C++ wrappers around a C header impl, and just write the C and C++ versions directly to the ABI, separately.  I think that will also reduce a lot of the confusion as you say about the moving pieces between core impl and wrappers.